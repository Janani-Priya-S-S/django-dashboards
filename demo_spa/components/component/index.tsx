export * from "./charts"
export * from "./html"
export * from "./table"
export * from "./text"
export * from "./display"