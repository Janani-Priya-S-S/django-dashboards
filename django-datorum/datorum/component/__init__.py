from .base import Component
from .chart import Chart
from .form import Form
from .map import Map
from .table import BasicTable, Table
from .text import CTA, Stat, Text


__all__ = [
    "Component",
    "CTA",
    "Chart",
    "Form",
    "Map",
    "Stat",
    "Text",
    "Table",
    "BasicTable",
]
