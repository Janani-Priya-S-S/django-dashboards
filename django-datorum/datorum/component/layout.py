import logging
from dataclasses import dataclass
from typing import Dict, Optional

from django.template import Context, Template
from django.template.loader import render_to_string
from django.utils.safestring import mark_safe


logger = logging.getLogger(__name__)


def css_template(width: int, css_classes: list):
    return f"span-{width} {css_classes}"


class LayoutBase:
    template_name: Optional[str] = None
    css_classes: Optional[str] = None
    width: Optional[int] = 6
    image_url: Optional[str] = None

    def __init__(
        self, *layout_components, css_classes=None, width=None, image_url=None
    ):
        if not layout_components:
            layout_components = []

        self.layout_components = layout_components

        if css_classes:
            self.css_classes = css_classes

        if width:
            self.width = width

        if image_url:
            self.image_url = image_url

    def get_components_rendered(self, dashboard, context: Dict) -> str:
        html = ""
        dashboard_components = dict([(x.key, x) for x in dashboard.get_components()])

        for layout_component in self.layout_components:
            dashboard_component = None

            # if str, we want the final dashboard_component
            if isinstance(layout_component, str):
                dashboard_component = dashboard_components.get(layout_component)

            if hasattr(layout_component, "render"):
                html += layout_component.render(dashboard, context)
            elif dashboard_component:
                html += str(dashboard_component)

        return mark_safe(html)


class ComponentLayout(LayoutBase):
    """
    holder for the components in the layout
    """

    def render(self, dashboard, context: Dict):
        return self.get_components_rendered(dashboard, context)


class HTMLComponentLayout(ComponentLayout):
    """
    template wrapper for components
    """

    def render(self, dashboard, context: Dict, **kwargs) -> str:
        components = self.get_components_rendered(dashboard, context)
        request = context.get("request")
        component_context = {
            "components": components,
            "css": css_template(self.width, self.css_classes),
        }

        return render_to_string(
            self.template_name, context=component_context, request=request
        )


class Card(HTMLComponentLayout):
    template_name: str = "datorum/layout/components/card.html"
    css_classes: str = "dashboard-component"


class Div(HTMLComponentLayout):
    template_name: str = "datorum/layout/components/div.html"


class TabContainer(HTMLComponentLayout):
    template_name: str = "datorum/layout/components/tabs/container.html"
    css_classes: str = "tab-content"

    def render(self, dashboard, context: Dict, **kwargs) -> str:
        tab_panels = self.get_components_rendered(dashboard, context)
        # make tab links for each tab
        links = "".join(tab.render_link() for tab in self.layout_components)

        request = context.get("request")
        component_context = {
            "css": css_template(self.width, self.css_classes),
            "links": links,
            "tab_panels": tab_panels,
            "first": self.layout_components[0],
        }

        html = render_to_string(
            self.template_name, context=component_context, request=request
        )
        return mark_safe(html)


class Tab(HTMLComponentLayout):
    tab_label: str = ""
    template_name: str = "datorum/layout/components/tabs/component.html"
    css_classes: str = "tab-pane fade"

    def render_link(self):
        return mark_safe(
            render_to_string(
                "datorum/layout/components/tabs/link.html",
                {
                    "link": self.tab_label,
                },
            )
        )

    def render(self, dashboard, context: Dict, **kwargs):
        return super().render(dashboard, context, **kwargs)


@dataclass
class HTML:
    html: str
    width: Optional[int] = 6

    def render(self, dashboard, context: Dict, **kwargs):
        to_render = f'<div class="{css_template(self.width, [])}">{self.html}</div>'
        return Template(to_render).render(Context(context))


@dataclass
class HR(HTML):
    html: str = "<hr />"


@dataclass
class Header(HTML):
    html: str = ""
    heading: str = ""
    size: int = 1

    def __post_init__(self):
        self.html = f"<h{self.size}>{self.heading}</h{self.size}>"
