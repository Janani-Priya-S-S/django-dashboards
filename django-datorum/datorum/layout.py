import logging
from typing import Dict

from django.template import Context, Template
from django.template.loader import render_to_string
from django.utils.safestring import mark_safe

logger = logging.getLogger(__name__)


class ComponentHolder:
    def __init__(self, *components):
        self.components = list(components)

    def get_components_rendered(self, dashboard, context: Dict):
        html = ""
        components = dict([(x.key, x) for x in dashboard.get_components()])

        for component in self.components:
            # layout component
            if hasattr(component, "render"):
                html += component.render(dashboard, context)
            else:
                # dashboard components
                try:
                    c = components[component]
                    html += str(c)
                except KeyError:
                    logger.warning(f"component {component} not found")
                    continue

        return mark_safe(html)


class LayoutComponent(ComponentHolder):
    """
    holder for the components in the layout
    """

    def render(self, dashboard, context: Dict):
        return self.get_components_rendered(dashboard, context)


class HTMLComponent(ComponentHolder):
    """
    template wrapper for components
    """

    template_name: str
    css_class_names: str = ""

    def __init__(self, *components, element_id: str, css_class_names: str = ""):
        self.element_id = element_id
        self.css_class_names += f" {css_class_names}"
        super().__init__(*components)

    def render(self, dashboard, context: Dict, **kwargs):
        components = self.get_components_rendered(dashboard, context)
        request = context.get("request")
        component_context = {
            "components": components,
            "css_class": self.css_class_names,
            "element_id": self.element_id,
        }

        return render_to_string(
            self.template_name, context=component_context, request=request
        )


class Card(HTMLComponent):
    template_name: str = "datorum/layout/components/card.html"
    css_class_names: str = "card"

    def __init__(
        self,
        *components,
        element_id: str,
        image_url: str = None,
        css_class_names: str = "",
    ):
        self.image_url = image_url
        super().__init__(
            *components, element_id=element_id, css_class_names=css_class_names
        )


class Div(HTMLComponent):
    template_name: str = "datorum/layout/components/div.html"
    css_class_names: str = "div"


class TabContainer(HTMLComponent):
    template_name: str = "datorum/layout/components/tabs/container.html"
    css_class_names: str = "tab-content"

    def render(self, dashboard, context: Dict, **kwargs):
        tab_panels = self.get_components_rendered(dashboard, context)
        # make tab links for each tab
        links = "".join(tab.render_link() for tab in self.components)

        request = context.get("request")
        component_context = {
            "css_class": self.css_class_names,
            "element_id": self.element_id,
            "links": links,
            "tab_panels": tab_panels,
            "first": self.components[0],
        }

        html = render_to_string(
            self.template_name, context=component_context, request=request
        )
        return mark_safe(html)


class Tab(HTMLComponent):
    template_name: str = "datorum/layout/components/tabs/component.html"
    css_class_names: str = "tab-pane fade"

    def __init__(
        self,
        tab_label,
        *components,
        element_id: str,
        css_class_names: str = "",
    ):
        super().__init__(
            *components, element_id=element_id, css_class_names=css_class_names
        )
        self.tab_label = tab_label

    def render_link(self):
        return mark_safe(
            render_to_string(
                "datorum/layout/components/tabs/link.html",
                {
                    "element_id": self.element_id,
                    "link": self.tab_label,
                },
            )
        )

    def render(self, dashboard, context: Dict, **kwargs):
        return super().render(dashboard, context, **kwargs)


class HTML:
    def __init__(self, html):
        self.html: str = html

    def render(self, dashboard, context: Dict, **kwargs):
        return Template(str(self.html)).render(Context(context))


class HR(HTML):
    def __init__(self):
        self.html = "<hr />"


class Header(HTML):
    def __init__(self, heading: str, size: int = 1):
        self.html = f"<h{size}>{heading}</h{size}>"
