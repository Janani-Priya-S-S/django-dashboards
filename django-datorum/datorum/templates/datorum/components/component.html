{% if component.is_deferred and htmx %}
    <div hx-get="{{ component.get_absolute_url }}?key={{ component.key }}" hx-trigger="intersect once"></div>
{% else %}
    {% include "datorum/components/filter.html" %}

    {# -inner will be replaced if a filter is applied. #}
    <div id="component-{{ component.key }}-inner" class="dashboard-component-inner fade-in">
        {% include component.template %}
        {% if component.cta %}
          <div>
            {% include component.cta.template with component=component.cta %}
          </div>
        {% endif %}
    </div>
{% endif %}