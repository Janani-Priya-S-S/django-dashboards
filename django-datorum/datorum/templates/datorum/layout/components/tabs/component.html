<div x-show="tab === '{{ element_id }}'">
    {% include 'datorum/layout/components/div.html' %}
</div>