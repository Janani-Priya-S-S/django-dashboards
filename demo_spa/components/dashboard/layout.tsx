import {LayoutComponentsWrapper} from "@/components";
import React from "react";
import {Dashboard} from "@/types";
import * as styles from './index.module.scss'

export const DashboardWithLayout: React.FC<{dashboard: Dashboard}> = ({dashboard}) => {
    return <div>
        <h1>{dashboard.Meta.name} (Layout applied)</h1>
        <div className={styles.dashboardContainer}>
            <LayoutComponentsWrapper dashboard={dashboard} layoutComponents={dashboard.Meta.layoutJson.layout_components}/>
        </div>
    </div>
}