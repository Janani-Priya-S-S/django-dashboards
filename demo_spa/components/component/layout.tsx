import React from "react";
import {
    HTMLComponent as HTMLComponentType,
    Dashboard,
    LayoutComponentTypes,
    LayoutComponent, DashboardComponentTypes,
} from "@/types";
import {HTML} from "@/components/component/html";
import * as styles from "@/components/component/index.module.scss";
import * as dashboardStyles from "@/components/dashboard/index.module.scss";
import {DashboardComponent} from "@/components/component/dashboard";


const LayoutComponentWrapperStyle = {
    [LayoutComponentTypes.Div]: styles.component,
    [LayoutComponentTypes.Card]: styles.component,
    [LayoutComponentTypes.TabContainer]: dashboardStyles.tabContainer,
    [LayoutComponentTypes.Tab]: styles.tab,
}

type HTMLComponentProps = {
    dashboard: Dashboard
    component: HTMLComponentType
}


export const HTMLComponent = ({dashboard, component}: HTMLComponentProps) => (
    <div className={dashboardStyles[`span${component.width}`]}>
        <HTML value={component.html}/>
    </div>
)


type LayoutComponentsWrapperProps = {
    dashboard: Dashboard
    layoutComponents: LayoutComponent[]
    visibleTab?: string
}


export const LayoutComponentsWrapper = ({dashboard, layoutComponents, visibleTab}: LayoutComponentsWrapperProps) => {
    return <>
        {layoutComponents.map((lc, i) => (
            (lc.renderType != LayoutComponentTypes.Tab || lc.tab_label == visibleTab) &&
            <LayoutComponentWrapper dashboard={dashboard} layoutComponent={lc} key={`wrapper${i}`}/>
        ))}
    </>
}


type ConditionalWrapperProps = {
    condition: boolean
    wrapper: (children: JSX.Element) => JSX.Element
    children: JSX.Element
}


const ConditionalWrapper = ({condition, wrapper, children}: ConditionalWrapperProps) =>
    condition ? wrapper(children) : children;


type LayoutComponentWrapperProps = {
    dashboard: Dashboard
    layoutComponent: LayoutComponent | HTMLComponentType | string | Object
}


export const LayoutComponentWrapper = ({dashboard, layoutComponent}: LayoutComponentWrapperProps) => {
    const isLayoutComponent = typeof layoutComponent === "object"
    const isTabContainer = (layoutComponent as LayoutComponent).renderType === LayoutComponentTypes.TabContainer

    let hasNested = false
    let castLayoutComponent: LayoutComponent

    if (isLayoutComponent) {
        castLayoutComponent = layoutComponent as LayoutComponent
        hasNested = isLayoutComponent && castLayoutComponent.layout_components?.length > 0
    }

    const [visibleTab, setVisibleTab] = React.useState(isTabContainer ? castLayoutComponent!.layout_components[0].tab_label : undefined)

    /*
        - Renders DashboardComponent when we are down to a string i.e the ref to put a component here.
        - else, Renders the outer LayoutComponent's class for style to be added
            - followed by nested LayoutComponentsWrapper if more nested layout_components are present
            - else, this is an HTML Component so render that.
    */
    return <>
        {isLayoutComponent && castLayoutComponent! ?
            <ConditionalWrapper
                condition={hasNested}
                wrapper={children => <div
                    className={`${LayoutComponentWrapperStyle[castLayoutComponent!.renderType]} ${dashboardStyles[`span${castLayoutComponent!.width || 6}`]}`}>{children}</div>}
            >
                <>
                    {
                        isTabContainer ?
                            <>
                                {castLayoutComponent.layout_components.map((tab, i) => {
                                    return <a className={dashboardStyles.span2}
                                              onClick={() => setVisibleTab(tab.tab_label)} key={`tab${i}`}>
                                        {tab.tab_label}
                                    </a>
                                })}
                                <LayoutComponentsWrapper dashboard={dashboard}
                                                         layoutComponents={castLayoutComponent.layout_components}
                                                         visibleTab={visibleTab}/>
                            </>
                            :
                            <>
                                {hasNested && castLayoutComponent ?
                                    <LayoutComponentsWrapper dashboard={dashboard}
                                                             layoutComponents={castLayoutComponent.layout_components}/>
                                    :
                                    <HTMLComponent dashboard={dashboard}
                                                   component={(layoutComponent as HTMLComponentType)}/>
                                }
                            </>
                    }
                </>
            </ConditionalWrapper> :
            <DashboardComponent dashboard={dashboard} component={dashboard.components.filter(c => c.key == layoutComponent)[0]}/>
        }
    </>
}
