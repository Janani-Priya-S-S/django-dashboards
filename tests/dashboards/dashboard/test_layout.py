from django.template import Context

import pytest

from tests.utils import render_dashboard_test
from wildcoeus.dashboards.component.layout import (
    HTML,
    Card,
    ComponentLayout,
    Div,
    Header,
    Tab,
    TabContainer,
)


pytest_plugins = [
    "tests.dashboards.fixtures",
]


def test_dashboard__render_layout(rf, dashboard_with_layout, snapshot):
    context = Context(
        {
            "dashboard": dashboard_with_layout,
            "request": rf.get("/"),
        }
    )

    snapshot.assert_match(render_dashboard_test(context))


@pytest.mark.parametrize("component_class", [Div, Tab, Card])
def test_html_component__render(rf, component_class, dashboard, snapshot):
    request = rf.get("/")
    context = Context({"request": request})

    snapshot.assert_match(
        component_class("component_1", css_classes="css_class").render(
            dashboard=dashboard(request=request), context=context
        )
    )


@pytest.mark.parametrize("component_class", [Div, Tab, Card])
def test_html_component__grid_css_classes__render(rf, component_class, dashboard):
    request = rf.get("/")
    context = Context({"request": request})

    html = component_class("component_1", grid_css_classes="grid-12").render(
        dashboard=dashboard(request=request), context=context
    )

    assert "grid-12" in html


@pytest.mark.django_db
@pytest.mark.parametrize("component_class", [Div, Tab, Card])
def test_html_component__grid_css_classes_not_set__default(
    rf, component_class, dashboard, settings
):
    settings.WILDCOEUS_DEFAULT_GRID_CSS = "default-css-class"

    request = rf.get("/")
    context = Context({"request": request})

    html = component_class("component_1").render(
        dashboard=dashboard(request=request), context=context
    )

    assert "default-css-class" in html


def test_tab_container__render(rf, dashboard, snapshot):
    request = rf.get("/")
    context = Context({"request": request})

    snapshot.assert_match(
        TabContainer(
            Tab(HTML("some text....")),
        ).render(dashboard=dashboard(request=request), context=context)
    )


def test_html__render(rf, dashboard, snapshot):
    context = Context({"request": rf.get("/")})

    snapshot.assert_match(
        HTML("some text....").render(dashboard=dashboard, context=context)
    )


def test_header__render(rf, dashboard, snapshot):
    context = Context({"request": rf.get("/")})

    snapshot.assert_match(
        Header(heading="some text....", size=2).render(
            dashboard=dashboard, context=context
        )
    )


def test_layout_component__render(rf, dashboard, snapshot):
    request = rf.get("/")
    context = Context({"request": request})

    snapshot.assert_match(
        ComponentLayout(
            "component_1",
            "component_2",
        ).render(dashboard=dashboard(request=request), context=context)
    )


def test_layout_component__unknown_component_ignored(rf, dashboard, snapshot):
    request = rf.get("/")
    context = Context({"request": request})

    snapshot.assert_match(
        ComponentLayout(
            "component_1",
            "component_2",
            "component_3",
        ).render(dashboard=dashboard(request=request), context=context)
    )
