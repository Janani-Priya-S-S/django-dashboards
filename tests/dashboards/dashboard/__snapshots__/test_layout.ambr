# serializer version: 1
# name: test_dashboard__render_layout
  '''
  }{<hr /><div class="span-6 ">
    <div class="span-6 ">
    
  
  
      
          <div id="component-dashapp1testdashboardcomponentcomponent_1-inner" class="dashboard-component-inner fade-in">
              
  value
  
          </div>
      
  
  
  
  </div><div class="span-6 ">
    
  
  
      <div hx-get="/dash/app1/testdashboard/@component/component_2/"
           hx-trigger="intersect once delay:1ms">
          <div class="htmx-indicator">
              
  <div class="loading-img">
      <img src="/static/dashboards/loading.svg" />
  </div>
          </div>
      </div>
  
  
  
  </div>
  </div>
  '''
# ---
# name: test_header__render
  '<h2>some text....</h2>'
# ---
# name: test_html__render
  'some text....'
# ---
# name: test_html_component__card__css_classes
  '''
  <div class="span-6">
    <div class="css_class">
      
      
      <div>
        
  
  
      
          <div id="component-dashapp1testdashboardcomponentcomponent_1-inner" class="dashboard-component-inner fade-in">
              
  value
  
          </div>
      
  
  
  
      </div>
      
    </div>
  </div>
  '''
# ---
# name: test_html_component__div__css_classes
  '''
  <div class="span-6 css_class">
    
  
  
      
          <div id="component-dashapp1testdashboardcomponentcomponent_1-inner" class="dashboard-component-inner fade-in">
              
  value
  
          </div>
      
  
  
  
  </div>
  '''
# ---
# name: test_html_component__render[Card]
  '''
  <div class="span-6">
    <div class="card">
      
      
      <div>
        
  
  
      
          <div id="component-dashapp1testdashboardcomponentcomponent_1-inner" class="dashboard-component-inner fade-in">
              
  value
  
          </div>
      
  
  
  
      </div>
      
    </div>
  </div>
  '''
# ---
# name: test_html_component__render[Div]
  '''
  <div class="span-6 ">
    
  
  
      
          <div id="component-dashapp1testdashboardcomponentcomponent_1-inner" class="dashboard-component-inner fade-in">
              
  value
  
          </div>
      
  
  
  
  </div>
  '''
# ---
# name: test_html_component__render[Tab]
  '''
  <div :class="{ ' active show': tab === 'component_1' }" x-show="tab === 'component_1'">
      
  </div>
  '''
# ---
# name: test_html_component__tab__css_classes
  '''
  <div :class="{ 'css_class active show': tab === 'component_1' }" x-show="tab === 'component_1'">
      
  </div>
  '''
# ---
# name: test_html_component__tabcontainer__css_classes
  '''
  <div class="span-6 css_class" x-data="{ tab: 'component_1' }">
      <div id="" class="tabs" >
          <div class="tab">
    <a :class="{ 'active': tab === 'component_1' }" x-on:click.prevent="tab = 'component_1'" href="#" class="">
      component_1
    </a>
  </div>
      </div>
      <hr>
      <div class="tab-content">
          <div :class="{ ' active show': tab === 'component_1' }" x-show="tab === 'component_1'">
      
  </div>
      </div>
  </div>
  '''
# ---
# name: test_layout_component__render
  '''
  
  
  
      
          <div id="component-dashapp1testdashboardcomponentcomponent_1-inner" class="dashboard-component-inner fade-in">
              
  value
  
          </div>
      
  
  
  
  
  
      <div hx-get="/dash/app1/testdashboard/@component/component_2/"
           hx-trigger="intersect once delay:1ms">
          <div class="htmx-indicator">
              
  <div class="loading-img">
      <img src="/static/dashboards/loading.svg" />
  </div>
          </div>
      </div>
  
  
  
  '''
# ---
# name: test_layout_component__unknown_component_ignored
  '''
  
  
  
      
          <div id="component-dashapp1testdashboardcomponentcomponent_1-inner" class="dashboard-component-inner fade-in">
              
  value
  
          </div>
      
  
  
  
  
  
      <div hx-get="/dash/app1/testdashboard/@component/component_2/"
           hx-trigger="intersect once delay:1ms">
          <div class="htmx-indicator">
              
  <div class="loading-img">
      <img src="/static/dashboards/loading.svg" />
  </div>
          </div>
      </div>
  
  
  
  
  
      
          <div id="component-dashapp1testdashboardcomponentcomponent_3-inner" class="dashboard-component-inner fade-in">
              
  value from callable
  
          </div>
      
  
  
  
  '''
# ---
# name: test_tab_container__render
  '''
  <div class="span-6 tab-container" x-data="{ tab: 'htmlhtmlsome-text' }">
      <div id="" class="tabs" >
          <div class="tab">
    <a :class="{ 'active': tab === 'htmlhtmlsome-text' }" x-on:click.prevent="tab = 'htmlhtmlsome-text'" href="#" class="">
      HTML(html=&#x27;some text....&#x27;)
    </a>
  </div>
      </div>
      <hr>
      <div class="tab-content">
          <div :class="{ ' active show': tab === 'htmlhtmlsome-text' }" x-show="tab === 'htmlhtmlsome-text'">
      
  </div>
      </div>
  </div>
  '''
# ---
